libpod/oci_util.go

```diff
@@ -112,19 +112,28 @@ func bindPorts(ports []ocicni.PortMapping) ([]*os.File, error) {
 func getOCIRuntimeError(runtimeMsg string) error {
 	includeFullOutput := logrus.GetLevel() == logrus.DebugLevel
 
+	runtimeMsg = strings.Trim(runtimeMsg, "\n")
+
 	if match := regexp.MustCompile("(?i).*permission denied.*|.*operation not permitted.*").FindString(runtimeMsg); match != "" {
 		errStr := match
 		if includeFullOutput {
 			errStr = runtimeMsg
 		}
-		return errors.Wrapf(define.ErrOCIRuntimePermissionDenied, "%s", strings.Trim(errStr, "\n"))
+
+		runtimeErr := errors.New(errStr)
+		return errors.Wrapf(runtimeErr, "%v", define.ErrOCIRuntimePermissionDenied)
```

Comment:
I'm against this - we want `define.ErrOCIRuntimePermissionDenied` to be the cause of the error, so we can check against it with `errors.Cause` later

Reply:
First of all, I'm a little disappointed that there isn't something like `errors.WrapWithError(cause, wrapper error)`. Since there isn't, I defined our own error type, `LibpodError`, that will respond to `errors.Is` appropriately.

```go
annotatedError := errors.Wrapf(runtimeErr, "%v", define.ErrOCIRuntime)
fmt.Println(annotatedError)                                       // OCI runtime error: writing file `cpu.cfs_quota_us`: Invalid argument
fmt.Println(errors.Cause(annotatedError))                         // writing file `cpu.cfs_quota_us`: Invalid argument
fmt.Println(errors.Cause(annotatedError) == define.ErrOCIRuntime) // false
fmt.Println(errors.Is(annotatedError, define.ErrOCIRuntime))      // true

moreAnnotatedError := errors.Wrap(annotatedError, "request failed")
fmt.Println(errors.Is(annotatedError, define.ErrOCIRuntime))      // true
```

---

I admit that I'm not super happy with [the implementation of LibpodErr.Is](https://github.com/xordspar0/podman/blob/04084a2c8e622d19664383e1f717168ac824aaa0/libpod/define/errors.go#L13-L15):

```go
func (e LibpodError) Is(target error) bool {
	return strings.HasPrefix(target.Error(), string(e))
}
```

Just checking for a string prefix is not very precise, but I couldn't figure out a way to insert the actual error into the chain. I'll keep thinking about this.

Reply:
So to be specific, I'm proposing that we use `errors.Is` instead of `errors.Cause` to identify what kind of error it is. Then it doesn't matter where in the chain the `define.ErrOCIRuntimePermissionDenied` error is.